prototypes/ScheduledMerges.hs

```diff
@@ -0,0 +1,551 @@
+
+                       -- let r = merge4 r1 r2 r3 r4
+                       --  in OngoingMerge 0 [r1, r2, r3, r4] r
+                     | OngoingMerge !MergeDebt ![Run] Run
+  deriving Show
+
+type Credit = Int
+type Debt   = Int
+
+type Run    = Map Key Op
+type Buffer = Map Key Op
+
+type Op     = Update Value Blob
+
+type Key    = Int
+type Value  = Int
+type Blob   = Int
+
+
+-- | The size of the 4 tiering runs at each level are allowed to be:
+-- @4^(level-1) < size <= 4^level@
+--
+tieringRunSize :: Int -> Int
+tieringRunSize n = 4^n
+
+-- | Levelling runs take up the whole level, so are 4x larger.
+--
+levellingRunSize :: Int -> Int
+levellingRunSize n = 4^(n+1)
+
+tieringRunSizeToLevel :: Run -> Int
+tieringRunSizeToLevel r =
+    1 + (finiteBitSize s - countLeadingZeros (s-1) - 1) `div` 2
+  where
+    s = Map.size r
+
+levellingRunSizeToLevel :: Run -> Int
+levellingRunSizeToLevel r = tieringRunSizeToLevel r - 1
+
+bufferSize :: Int
+bufferSize = tieringRunSize 1 -- 4
+
+mergePolicyForLevel :: Int -> [Level s] -> MergePolicy
+mergePolicyForLevel 1 [] = MergePolicyTiering
+mergePolicyForLevel _ [] = MergePolicyLevelling
+mergePolicyForLevel _ _  = MergePolicyTiering
+
+mergeLastForLevel :: [Level s] -> MergeLastLevel
+mergeLastForLevel [] = MergeLastLevel
+mergeLastForLevel _  = MergeMidLevel
+
+invariant :: forall s. HasCallStack => Levels s -> ST s Bool
+invariant = go 1
+  where
+    go :: Int -> [Level s] -> ST s Bool
+    go !_ []     = return True
+
+    go !ln (Level mr rs : ls) = do
+
+      mrs <- case mr of
+               SingleRun r        -> return (CompletedMerge r)
+               MergingRun _ _ ref -> readSTRef ref
+
+      assert (case mr of
+                SingleRun{} -> True
+                MergingRun mp ml _ -> mergePolicyForLevel ln ls == mp
+                                   && mergeLastForLevel ls == ml)
+        assert (length rs <= 3) $
+        assert (expectedRunLengths ln rs ls) $
+        assert (expectedMergingRunLengths ln mr mrs ls) $
+        return ()
+
+      go (ln+1) ls
+
+    -- All runs within a level "proper" (as opposed to the incoming runs
+    -- being merged) should be of the correct size for the level.
+    expectedRunLengths :: Int -> [Run] -> [Level s] -> Bool
+    expectedRunLengths ln rs ls =
+      case mergePolicyForLevel ln ls of
+        MergePolicyLevelling -> null rs
+        MergePolicyTiering   -> all (\r -> tieringRunSizeToLevel r == ln) rs
+
+    -- Incoming runs being merged also need to be of the right size, but the
+    -- conditions are more complicated.
+    expectedMergingRunLengths :: Int -> MergingRun s -> MergingRunState
+                              -> [Level s] -> Bool
+    expectedMergingRunLengths ln mr mrs ls =
+      case mergePolicyForLevel ln ls of
+        MergePolicyLevelling ->
+          case (mr, mrs) of
+            -- A single incoming run (which thus didn't need merging) must be
+            -- of the expected size already
```

Comment:
Yes. I'll change that to size range. Also, in later work (not in this PR) I've updated the code to get rid of `levellingRunSizeToLevel` and only use size ranges. Using size ranges is easier to generalise to variable powers (rather than hard-coded to 4).